File: samples/kotlin/src/main/kotlin/com/grid/sample/Log.kt
package com.grid.sample

import org.slf4j.LoggerFactory

object Log {
    private val logger = LoggerFactory.getLogger("grid-sample")

    private const val RESET = "\u001B[0m"
    private const val CYAN = "\u001B[36m"
    private const val GREEN = "\u001B[32m"
    private const val YELLOW = "\u001B[33m"
    private const val MAGENTA = "\u001B[35m"
    private const val RED = "\u001B[31m"
    private const val DIM = "\u001B[2m"
    private const val BOLD = "\u001B[1m"

    fun incoming(method: String, path: String, body: String? = null) {
        logger.info("${CYAN}${BOLD}⬇ FE REQUEST${RESET}  ${CYAN}$method $path${RESET}")
        body?.let {
            if (it.isNotBlank()) logger.info("${DIM}  Body: ${prettyOneLine(it)}${RESET}")
        }
    }

    fun gridRequest(operation: String, detail: String? = null) {
        logger.info("${YELLOW}${BOLD}➡ GRID API${RESET}   ${YELLOW}$operation${RESET}")
        detail?.let {
            logger.info("${DIM}  Params: ${prettyOneLine(it)}${RESET}")
        }
    }

    fun gridResponse(operation: String, response: String) {
        logger.info("${GREEN}${BOLD}⬅ GRID RESP${RESET}  ${GREEN}$operation${RESET}")
        logger.info("${DIM}  ${prettyOneLine(response)}${RESET}")
    }

    fun gridError(operation: String, error: Exception) {
        logger.error("${RED}${BOLD}✘ GRID ERR${RESET}   ${RED}$operation: ${error.message}${RESET}")
    }

    fun webhook(body: String) {
        logger.info("${MAGENTA}${BOLD}⚡ WEBHOOK${RESET}   ${MAGENTA}${prettyOneLine(body)}${RESET}")
    }

    fun webhookSignatureResult(valid: Boolean) {
        if (valid) {
            logger.info("${GREEN}  Signature: valid${RESET}")
        } else {
            logger.warn("${RED}  Signature: INVALID${RESET}")
        }
    }

    fun response(status: Int) {
        val color = if (status < 400) GREEN else RED
        logger.info("${color}${BOLD}⬆ FE RESPONSE${RESET} ${color}$status${RESET}")
    }

    private fun prettyOneLine(json: String): String {
        val trimmed = json.trim()
        return if (trimmed.length > 500) trimmed.take(500) + "..." else trimmed
    }
}

File: samples/typescript/src/log.ts
const RESET = "\x1b[0m";
const CYAN = "\x1b[36m";
const GREEN = "\x1b[32m";
const YELLOW = "\x1b[33m";
const MAGENTA = "\x1b[35m";
const RED = "\x1b[31m";
const DIM = "\x1b[2m";
const BOLD = "\x1b[1m";

function truncate(str: string, max = 500): string {
  const trimmed = str.trim();
  return trimmed.length > max ? trimmed.slice(0, max) + "..." : trimmed;
}

export const log = {
  incoming(method: string, path: string, body?: string) {
    console.log(`${CYAN}${BOLD}⬇ FE REQUEST${RESET}  ${CYAN}${method} ${path}${RESET}`);
    if (body) console.log(`${DIM}  Body: ${truncate(body)}${RESET}`);
  },

  gridRequest(operation: string, detail?: string) {
    console.log(`${YELLOW}${BOLD}➡ GRID API${RESET}   ${YELLOW}${operation}${RESET}`);
    if (detail) console.log(`${DIM}  Params: ${truncate(detail)}${RESET}`);
  },

  gridResponse(operation: string, response: string) {
    console.log(`${GREEN}${BOLD}⬅ GRID RESP${RESET}  ${GREEN}${operation}${RESET}`);
    console.log(`${DIM}  ${truncate(response)}${RESET}`);
  },

  gridError(operation: string, error: unknown) {
    const msg = error instanceof Error ? error.message : String(error);
    console.error(`${RED}${BOLD}✘ GRID ERR${RESET}   ${RED}${operation}: ${msg}${RESET}`);
  },

  webhook(body: string) {
    console.log(`${MAGENTA}${BOLD}⚡ WEBHOOK${RESET}   ${MAGENTA}${truncate(body)}${RESET}`);
  },

  response(status: number) {
    const color = status < 400 ? GREEN : RED;
    console.log(`${color}${BOLD}⬆ FE RESPONSE${RESET} ${color}${status}${RESET}`);
  },
};
